feat(posts): the Go benchmark that measured nothing (series 1/5) (#312)
kakkoyun/me: content/posts/go-benchmarks-lying-compiler-honesty.md

---
title: "The Go Benchmark That Measured Nothing: Compiler Honesty in testing.B"
description: "Dead-code elimination, constant folding, and inlining can silently gut a Go benchmark loop. Learn how to detect each transformation, why allocs/op is the honest signal, and how testing.B.Loop in Go 1.24 removes most of these footguns at the language level."

> date: 2026-09-01T00:00:00Z
> publishDate: 2026-09-01T00:00:00Z

Author: New slot: **Friday, August 21** (moved from Tue Sep 1). The series moves as a block to consecutive Fridays: 1/5 Aug 21, 2/5 Aug 28, 3/5 Sep 4, 4/5 Sep 11, 5/5 Sep 18.

```suggestion
date: 2026-08-21T00:00:00Z
publishDate: 2026-08-21T00:00:00Z
```

Why Friday and why earlier: Fridays are the only day the promotion cron fires, and a five-part technical series is the queue's best promotion material. Tuesdays are committed to the agentic-coding series, which you've opted out of promotion — so it was either move this series to Fridays or run it with no promotion at all. Moving it up eleven days is a side effect of taking the Friday slots, and it's affordable here because these five have correct frontmatter and an active revision loop, which the keynote drop in #311 doesn't yet.

None of the five carries a `promote:` key, so all five get promoted. That's five consecutive weeks of the promotion slot going to one series — worth knowing before you agree to it.

---
_Generated by [Claude Code](https://claude.ai/code)_

> date: 2026-09-01T00:00:00Z
> publishDate: 2026-09-01T00:00:00Z

Author: Supersedes my earlier Aug 21 suggestion — that one hasn't been applied, so this replaces it cleanly. **Fri Aug 28**, one week later, freeing Aug 21 for #205.

```suggestion
date: 2026-08-28T00:00:00Z
publishDate: 2026-08-28T00:00:00Z
```

The series still gets every part promoted, still runs on consecutive Fridays, and still starts earlier than the Tue Sep 1 it had yesterday.

---
_Generated by [Claude Code](https://claude.ai/code)_

> draft: true

Author: This is the only thing standing between the post and tomorrow's build. With it set, the Sep 1 cron builds and ships nothing.

```suggestion
```

The dates above are already correct and need no change — `date` and `publishDate` both read 2026-09-01, so the post will display tomorrow's date on the day it goes live.

---
_Generated by [Claude Code](https://claude.ai/code)_

A single number is not a result. It is one sample from a distribution. Two benchmarks can share a mean and have completely different distributions. Post 2, [A Single Benchmark Number Is a Lie](/posts/go-benchmarks-lying-statistics/), covers how to read that distribution, how to use `benchstat` to compare results across commits, and when a measured difference is real rather than noise.

---

### Resources

- Demo code: [`talks/go-benchmarks-lying/demo/`](https://github.com/kakkoyun/gopherconuk-26/tree/main/talks/go-benchmarks-lying/demo) — `make bench-dce`, `make asm-dce`, `make bench-timer`
- Talk: "Why Your Go Benchmarks Are Lying (And How to Stop Them)", GopherCon UK 2026
- Prerequisite post: [Measuring Software Performance: Why Your Benchmarks Are Probably Lying](/posts/fosdem-2026-measuring-software-performance/)
- Dave Cheney — [How to write benchmarks in Go](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go)

> - Go Team — [Evolving the Go benchmark API](https://go.dev/blog/testing-b-loop) (the `testing.B.Loop` announcement post)

Author: Wrong title — the Go blog post at this URL is **"More predictable benchmarking with testing.B.Loop"** (verified against the live page just now). "Evolving the Go benchmark API" isn't its title at any point in its history I can find.

```suggestion
- Go Team — [More predictable benchmarking with testing.B.Loop](https://go.dev/blog/testing-b-loop) (the `testing.B.Loop` announcement post)
```

---
_Generated by [Claude Code](https://claude.ai/code)_
